src/NLog.Web.AspNetCore/NLogRequestLoggingMiddleware.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace NLog.Web
+{
+    /// <summary>
+    /// Middleware that writes all requests to Logger named "RequestLogging"
+    /// </summary>
+    /// <remarks>
+    /// - LogLevel.Error - Request failed with exception<br/>
+    /// - LogLevel.Warn  - Request completed with StatusCode >= 400<br/>
+    /// - LogLevel.Info  - Request completed normal StatusCode<br/>
+    /// </remarks>
+    public class NLogRequestLoggingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly Microsoft.Extensions.Logging.ILogger _logger;
+
+        /// <summary>
+        /// Initializes new instance of the <see cref="NLogRequestLoggingMiddleware"/> class
+        /// </summary>
+        /// <remarks>
+        /// Use the following in Startup.cs:
+        /// <code>
+        /// public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+        /// {
+        ///    app.UseMiddleware&lt;NLog.Web.RequestLoggingMiddleware&gt;();
+        /// }
+        /// </code>
+        /// </remarks>
+        public NLogRequestLoggingMiddleware(RequestDelegate next, ILoggerFactory loggerFactory)
+        {
+            _next = next;
+            _logger = loggerFactory?.CreateLogger("RequestLogging") ?? Microsoft.Extensions.Logging.Abstractions.NullLogger.Instance;
+        }
+
+        /// <summary>
+        /// Executes the middleware.
+        /// </summary>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            try
+            {
+                await _next(httpContext);
+                LogHttpRequest(httpContext, null);
+            }
+            catch (Exception exception) when (LogHttpRequest(httpContext, exception))
```

Comment:
this is bit confusing, the `when` condition has a side effect?

Reply:
is this intended?

Reply:
Yes this is the new Exception-filter-logic, that allows one to react to the first-chance exception (keeping scope state). Instead of handling it in the normal exception-catch where one have left the scope. See also: https://docs.microsoft.com/en-us/dotnet/standard/exceptions/using-user-filtered-exception-handlers

Reply:
From my first view, I would say that this is the same:

```c#
catch (Exception exception){
LogHttpRequest(httpContext, exception)
}

So the difference is the scope? It would be great if we could ensure this from the tests.

Please document in code that we do this "trick" for the scope :)
```